bootstrap/kubeadm/internal/cloudinit/controlplane_join.go

```diff
@@ -18,11 +18,137 @@ package cloudinit
 
 import (
 	"github.com/pkg/errors"
-
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3"
 	"sigs.k8s.io/cluster-api/util/secret"
 )
 
 const (
+	controlPlaneJoinScript = `#!/bin/bash
+# Copyright 2020 The Kubernetes Authors.
```

Comment:
I like that we are adding retries to the kubeadm join, it's a good step towards making cluster api bootstrap more resilient. 

One concern I have with this PR however is that we are significantly expanding the size of the join bash script passed via cloud init. 

1) Both AWS and Azure (and others too I'm sure) have strict size limits on instance/VM user data (for AWS it's 16 KB https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-add-user-data.html, for Azure it's 64KB https://docs.microsoft.com/en-us/azure/virtual-machines/custom-data#passing-custom-data-to-the-vm). We need to plan now rather than later how we're going to keep the size manageable as we add more features, configurations and resiliency kubeadm bootstrapping. Those are pretty small limits and we'll get there faster than we think (based on previous experience doing Kubernetes bootstrapping with cloud init scripts).

2) How are we going to keep this bash script maintainable, readable, and well tested now that we are embedding logic inside of it? Before this change, there was only one line of bash in this file which made it easy to maintain, but now it is opening the door to more logic and functions being added in here in the future. It will quickly become unmanageable (IMO) if we keep that logic as a const in a go file.

Reply:
I have very much the same concerns, which is why I had closed it.

For AWS, the fact that we actually store this script in 10KB chunks in AWS Secrets Manager means CAPA effectively now supports arbitrary length userdata. Azure is a valid concern however.

Long term, I think we should get more retries into kubeadm itself.

Reply:
Scripts in variable in golang are scary and unpractical. You can't lint them properly, and the only way to run them is via the go that embeds them. 
If keeping that direction, I would try having them as `*.sh` files and embed them as resources at compile time?

What about writing go instead? move all the scripted logic into the binary. The cloudinit can then install the binary as a systemd unit (if not present on the image) and run it.

kubeadm for example can be used as a library in the binary.

Reply:
We can't use kubeadm as a library yet as it's in-tree in kubernetes.

Reply:
These are all super valid concerns, thanks for bringing them up! We've noticed multiple `kubeadm join` failures in different environments and that's how we arrived at this PR.

I'd propose the following to this group:
- We'll keep any added retry functionality under a feature gate which is off by default, or in alternative an option on the bootstrap KubeadmConfig.
- Let's gather data and make sure that this fixes the underlying issues we're seeing, meanwhile we try to work towards a better solution.
- We'll put some work and retries in kubeadm as well and adopt it once we raise the minimum Kubernetes versions for workload clusters, or in alternative code in kubeadm bootstrap controller can decide what do to depending on the k8s version specified (this will require to know which patch version in kubeadm/k8s implemented retries).

For the current PR, we should do whatever we can do keep the amount of increase in size to a minimum, so let's remove anything that's not required.

what do you all think?

Reply:
It should also be noted that everything in this PR is a temporary workaround and not a long term solution.

Reply:
> We'll keep any added retry functionality under a feature gate which is off by default, or in alternative an option on the bootstrap KubeadmConfig.

+1

> Let's gather data and make sure that this fixes the underlying issues we're seeing, meanwhile we try to work towards a better solution.

this is the part i'm interested about. we seem to be applying a workaround over a LB setup issue. we need answers on this part that @randomvariable mentioned:
>> HAProxy always starts new backends as ready, pending healthchecks (if anyone knows how to change that I'm all ears)

sending a support ticket for HAProxy might be an option.

> We'll put some work and retries in kubeadm as well and adopt it once we raise the minimum Kubernetes versions for workload clusters, or in alternative code in kubeadm bootstrap controller can decide what do to depending on the k8s version specified (this will require to know which patch version in kubeadm/k8s implemented retries).

mostly blocked on developer bandwidth. we are aware that kubeadm is not retrying for all API calls, but there are a lot of cases and we need to make sense of where this needs to apply. a kubeadm join is currently a best effort that assumes that the user infra is setup and ready when the join command is called.

Reply:
The plan outlined by @vincepri above sounds good to me. Can we capture "meanwhile we try to work towards a better solution" in an issue?

> For the current PR, we should do whatever we can do keep the amount of increase in size to a minimum, so let's remove anything that's not required.

+1 on this

Reply:
long-term plan outlined here https://github.com/kubernetes-sigs/cluster-api/issues/2769